elide-core/src/main/java/com/yahoo/elide/core/filter/Predicate.java

```diff
@@ -83,4 +74,87 @@
 
         return predicates;
     }
+
+    /**
+     * Retrieve the type of the object to be filtered.
+     *
+     * NOTE: Types are now fully qualified by (ROOT_TYPE.FIELD1.FIELD2..FIELD(N-1)).ACCESSED_FIELD
+     *
+     * @param types Array of types
+     * @return Type of object to be filtered. Empty string if nothing found.
+     */
+    private static String getType(final String[] types) {
+        if (types == null || types.length <= 1) {
+            return "";
+        }
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0 ; i < types.length - 1 ; ++i) {
+            sb.append(types[i]);
+            if (i != types.length - 2) {
+                sb.append(".");
+            }
+        }
+        return sb.toString();
+    }
+
+    /**
+     * Retrieve the type of the field being accessed.
+     *
+     * @param types Array of types
+     * @return Type of the field
+     */
+    private static Class<?> getFieldType(final Class<?>[] types) {
+        if (types == null || types.length <= 1) {
+            throw new InvalidPredicateException("Unknown type for field");
+        }
+        return types[types.length - 1];
+    }
+
+    /**
+     * Get the classes of each field contained within the key part.
+     *
+     * NOTE: This method checks that the specified traversal path to a particular type
+     *       is valid. In the case of an invalid field, it will throw an "InvalidPredicateException."
+     *
+     * @param keyParts Key components
+     * @param dictionary Entity dictionary
+     * @return An array containing types for each of the key parts (in order). Empty array if none found.
+     */
+    private static Class<?>[] getTypes(final String[] keyParts, final EntityDictionary dictionary) {
+        if (keyParts == null || keyParts.length <= 0) {
+            return new Class[0];
+        }
+        Class<?>[] types = new Class[keyParts.length];
+        String type = keyParts[0];
+        types[0] = dictionary.getBinding(type);
+        for (int i = 1 ; i < keyParts.length ; ++i) {
+            final String field = keyParts[i];
+            final Class<?> entityClass = types[i - 1];
+            if (entityClass == null) {
```

Comment:
As far as I can tell, this should only ever happen on types[0], because you check for all other possible ones in line 140. So maybe you should move this check after `types[0] = dictionary.getBinding(type);`